### ISSUE TITLE
How to Configure deploy widget ( Udeploy Collector issue)

### ISSUE THREAD

**Issue by shilpsl1** [Hygieia/Hygieia#1084]
Hi,
 I am unable to see anything in deploy widget of Hygieia .
I imported the SSL cert and was able to run collector successfully. But I donot see proper data in hygieia.  Below are api output and screen shots attached.
I see only application name rest are empty. Below are results

    https://10.0.0.0:8443/rest/deploy/application
    [{"id":"8db06d0d-874f-4338-9903-b3de34f2de05","securityResourceId":"35981c95-7851-4636-921b-938e5b2435ad","name":"Sample_App_Deploy","description":"","created":1483992118553,"enforceCompleteSnapshots":false,"active":true,"tags":[],"deleted":false,"user":"admin"}]

    https://10.0.0.0:8443/rest/deploy/application/8db06d0d-874f-4338-9903-b3de34f2de05/environments/false
    [{"security":{"Create Environments":true,"Create Environments From Template":true,"Delete":true,"write":true,"Edit Basic Settings":true,"execute":true,"Execute on Environments":true,"Manage Approval Processes":true,"Manage Base Resources":true,"Manage Properties":true,"Manage Teams":true,"read":true,"View Environments":true},"id":"71a8834b-0036-4e27-a57e-d8b9339fb325","securityResourceId":"a9e0a5f3-7ec3-4061-9d4d-2dceec475140","name":"Test_Env","description":"","color":"#00B2EF","requireApprovals":false,"noSelfApprovals":false,"lockSnapshots":false,"calendarId":"5326a8cf-30e5-42c5-ab2b-600c8b44f54e","active":true,"deleted":false,"cleanupDaysToKeep":0,"cleanupCountToKeep":0,"enableProcessHistoryCleanup":false,"useSystemDefaultDays":true,"historyCleanupDaysToKeep":365,"conditions":[]}]

    https://10.0.0.0:8443/rest/deploy/environment/71a8834b-0036-4e27-a57e-d8b9339fb325/resources
    [{"security":{"Create Resources":true,"Delete":true,"write":true,"Edit Basic Settings":true,"Manage Children":true,"Manage Impersonation":true,"Manage Properties":true,"Manage Teams":true,"Map to Environments":true,"read":true,"View Resources":true},"id":"6659ade4-9915-4a36-a71a-bf2810e58dc5","securityResourceId":"17118bda-f1c9-4645-a2ed-d3cf7838bc14","name":"Sample_App_Deploy","path":"/Sample_App_Deploy","active":true,"description":"","inheritTeam":false,"discoveryFailed":false,"prototype":false,"impersonationPassword":"****","impersonationUseSudo":false,"impersonationForce":false,"hasAgent":false,"tags":[],"resource":{"id":"17118bda-f1c9-4645-a2ed-d3cf7838bc14"},"versions":[],"hasChildren":true,"isRoot":true}]

    https://10.0.0.0:8443/rest/deploy/environment/71a8834b-0036-4e27-a57e-d8b9339fb325/noncompliantResources
    [ ]

**Comment by shilpsl1** [comment]
![4024809edcf05957](https://cloud.githubusercontent.com/assets/24657895/21861092/6660eaf0-d801-11e6-9b1f-32e8053957cd.jpg)
![1c7756425028ab52](https://cloud.githubusercontent.com/assets/24657895/21861101/6b88ae82-d801-11e6-86e8-7cded5e9f0ad.jpg)

**Comment by nireeshT** [comment]
@shilpsl1  - I see the rest call is success returning environment details. Check if collector task is insert/updating values into MongoDB.

**Comment by shilpsl1** [comment]
**I see in DB that Sample_App_Deploy entry is present**

> db.collectors.find()
{ "_id" : ObjectId("585af60a0b2f38094d7800c3"), "_class" : "com.capitalone.dashboard.model.Collector", "name" : "Product", "collectorType" : "Product", "enabled" : true, "online" : true, "lastExecuted" : NumberLong(0) }
{ "_id" : ObjectId("5873e83c0b2f38726da5a067"), "_class" : "com.capitalone.dashboard.model.UDeployCollector", "udeployServers" : [ "https://10.0.0.0:8443" ], "niceNames" : [ ], "name" : "UDeploy", "collectorType" : "Deployment", "enabled" : true, "online" : true, "lastExecuted" : NumberLong("1484169300192") }


> db.components.find()
{ "_id" : ObjectId("587002500b2f387f4f0aee71"), "_class" : "com.capitalone.dashboard.model.Component", "name" : "DevOps", "collectorItems" : { "Deployment" : [ { "_id" : ObjectId("587533e00b2f3858c0d4410c"), "description" : "Sample_App_Deploy", "enabled" : true, "pushed" : false, "collectorId" : ObjectId("5873e83c0b2f38726da5a067"), "lastUpdated" : NumberLong(0), "options" : { "applicationId" : "8db06d0d-874f-4338-9903-b3de34f2de05", "applicationName" : "Sample_App_Deploy", "instanceUrl" : "https://10.0.0.0:8443" } } ] } }
{ "_id" : ObjectId("5873e8170b2f384502062b19"), "_class" : "com.capitalone.dashboard.model.Component", "name" : "DevOps Dash", "collectorItems" : { "Product" : [ { "_id" : ObjectId("587002500b2f387f4f0aee73"), "description" : "DevOps Dashboard", "enabled" : true, "pushed" : false, "collectorId" : ObjectId("585af60a0b2f38094d7800c3"), "lastUpdated" : NumberLong(0), "options" : { "dashboardId" : "587002500b2f387f4f0aee72" } } ] } }

**Comment by nireeshT** [comment]
@shilpsl1  - how about --- db.getCollection('environment_components').find({})

**Comment by shilpsl1** [comment]
It does not show any output. Empty

**Comment by nireeshT** [comment]
@shilpsl1  - UDeploy collector is not updating env info into MongoDB. Check properties for db user and pwd are valid and test connection. Follow https://github.com/capitalone/Hygieia/blob/master/collectors/deploy/udeploy/README.md for sample properties. if issues is not resolved,please attach complete UDeploy Collector log.

**Comment by shilpsl1** [comment]
Thanks for the reply...
DB connection works fine.. Please check the attached log file, property file and jar file which I have tried.

I have tried running all possibilities as shown below,
cd Hygieia-master/collectors/deploy/udeploy/target

java -jar udeploy-deployment-collector-2.0.4-SNAPSHOT.jar --spring.config.name=application –spring.config.location=application.properties

java -jar udeploy-collector-0.1.1.jar --spring.config.name=application –spring.config.location=application.properties

[application.properties.txt](https://github.com/capitalone/Hygieia/files/702364/application.properties.txt)
[deploycollector-2017-01-12.0.log.txt](https://github.com/capitalone/Hygieia/files/702365/deploycollector-2017-01-12.0.log.txt)
[udeploy-collector-0.1.1.jar.zip](https://github.com/capitalone/Hygieia/files/702367/udeploy-collector-0.1.1.jar.zip)

**Comment by nireeshT** [comment]
@shilpsl1 - Is Sample_app_Deploy actually deployed? which browser is used?

**Comment by haquefire** [comment]
Not sure if this helps but I am experiencing similar behavior.  I am running my UDeploy Collector through a debugger and noticing that environment variable in UDeployCollectorTask.java is being populated properly wiht my environment "dev".   But compList and combinedDataList are of size 0.   The calls to the uDeploy rest api for my environment seem to be working fine.   

Looking in mongo I see  my applicationName being populated in the collector_items, but I don't see anything related to environments being populated anywhere in the db.

**Comment by shilpsl1** [comment]
Am I missing any step ??? I am not sure what is going wrong ...

**Comment by shouebs** [comment]
I have been facing similar problem. I have my api and uDeploy collector running but i dont see any output on the Deploy widget.

2017-01-17 14:54:59,998 INFO  c.c.d.collector.CollectorTask - Running Collector: UDeploy
2017-01-17 14:55:00,000 INFO  c.c.d.collector.CollectorTask - -----------------------------------
2017-01-17 14:55:00,001 INFO  c.c.d.collector.CollectorTask - http://mycompanyud:8080
2017-01-17 14:55:00,001 INFO  c.c.d.collector.CollectorTask - -----------------------------------
2017-01-17 14:55:00,029 INFO  c.c.d.collector.CollectorTask - All apps                2        0s
2017-01-17 14:55:00,035 INFO  c.c.d.collector.CollectorTask - New apps                0        0s
2017-01-17 14:55:00,090 INFO  c.c.d.collector.CollectorTask -  XLR-Test                        0s
2017-01-17 14:55:00,091 INFO  c.c.d.collector.CollectorTask - Finished                         0s

Mongo DB values:

**collectors**
> db.collectors.find()
{ "_id" : ObjectId("587e657810f90506d0a9af6c"), "_class" : "com.capitalone.dashboard.model.UDeployCollector", "udeployServers" : [ "http://mycompanyud:8080" ], "niceNames" : [ ]
, "name" : "UDeploy", "collectorType" : "Deployment", "enabled" : true, "online" : true, "lastExecuted" : NumberLong("1484683200099") }

**collector_items**
> db.collector_items.find()
{ "_id" : ObjectId("587e657810f90506d0a9af6e"), "_class" : "com.capitalone.dashboard.model.UDeployApplication", "description" : "XLR-Test", "enabled" : true, "pushed" : false, "col
lectorId" : ObjectId("587e657810f90506d0a9af6c"), "lastUpdated" : NumberLong(0), "options" : { "applicationId" : "e8f239f7-c8d1-445d-bfb5-90e01aa13c2e", "applicationName" : "XLR-Te
st", "instanceUrl" : "http://mycompanyud:8080" } }


I have also attahced the json output for application, environment false and resources. The noncompliant resource is null.

**Comment by spyndi** [comment]
@shilpsl1 /@shouebs 

Any luck/progress on this issue? I also see the same behavior with the UDeploy collector. I was about to start debugging/troubleshooting the issue, but wanted to check if you have found a solution.

**Comment by shilpsl1** [comment]
@spyndi 
Tried all possibility, nothing seems to work so far ..

**Comment by tabladrum** [comment]
@spyndi @shilpsl1 when was the last time you ran the deployment job in UCD?

**Comment by shilpsl1** [comment]
I try daily, even yesterday made a deployment to check whether it works.
UCD Version which i am using -  6.2.1.

**Comment by shouebs** [comment]
@tabladrum, same here, there was a deployment yesterday. the version of the UCD is 6.2.2.

**Comment by shouebs** [comment]
@tabladrum @shilpsl1 @spyndi 

Is there any progress on this issue ? Did anyone get it working ?

**Comment by shilpsl1** [comment]
we had a call and found that i was using 6.2.1 version of UCD. It is working fine with UCD version 6.1. 
May be some issue with 6.2 and above version of UCD.
@shouebs  what version of UCD are you using ?

**Comment by shouebs** [comment]
@shilpsl1  Thanks for the update. I am using 6.2.2 .  Do you have any information if the issue with 6.2 will be fixed ?

**Comment by shilpsl1** [comment]
@shouebs 
May be @tabladrum has to answer this... He said he would check and come back to me this week on this issue. But I did hear anything so far from him ....

**Comment by shouebs** [comment]
@shilpsl1 Thanks.
@tabladrum Please keep us posted if there is any progress with fixing the issue with UCD version 6.2 and above.

**Comment by spyndi** [comment]
@shilpsl1 @tabladrum 
FYI - We are also using UCD 6.2.1 version and running into the same issue as reported above.

@tabladrum - I was doing some debugging of the UDeploy collector code and may have some additional data for troubleshooting.

I see that the method "getLowestLevelChildren" in DefaultUDeployClient class is returning a null JSONArray and hence the subsequent calls in the stack is not collecting any data for the widget.

`private JSONArray getLowestLevelChildren(JSONObject topParent, JSONArray returnArray) {
        JSONArray jsonChildren = (JSONArray) topParent.get("children");

        if (jsonChildren != null && jsonChildren.size() > 0) {
            for (Object child : jsonChildren) {
                if (!hasChildren((JSONObject) child)) {
                    returnArray.add(child);
                } else {
                    getLowestLevelChildren((JSONObject) child, returnArray);
                }
            }
        }
        return returnArray;
    }`

The input JSONObject to this method is -

{"parent":{"inheritTeam":false,"active":true,"description":"Test - Application","prototype":false,"tags":[],"impersonationForce":false,"path":"\/test-103831","security":{"Delete":false,"Manage Impersonation":false,"Manage Teams":false,"Edit Basic Settings":false,"read":false,"Map to Environments":false,"View Resources":false,"write":false,"Manage Children":false,"Create Resources":false,"Manage Properties":false},"hasAgent":false,"name":"test-103831","impersonationPassword":"****","securityResourceId":"d346ec32-9622-4364-92d2-4938617ca6bc","id":"f3c72817-394e-48f5-8e4a-f314816a61f6","impersonationUseSudo":false,"discoveryFailed":false},"isRoot":true,"inheritTeam":true,"resource":{"id":"c0d838a4-fe9a-4533-a0b1-33c87629a701"},"hasChildren":true,"active":true,"description":"CIT2","type":"subresource","prototype":false,"tags":[],"impersonationForce":false,"path":"\/test-103831\/CIT2","security":{"Delete":false,"Manage Impersonation":false,"Manage Teams":false,"Edit Basic Settings":false,"read":true,"Map to Environments":false,"View Resources":true,"write":false,"Manage Children":false,"Create Resources":false,"Manage Properties":false},"hasAgent":false,"versions":[],"name":"CIT2","impersonationPassword":"****","securityResourceId":"c0d838a4-fe9a-4533-a0b1-33c87629a701","id":"301b972c-668e-4b9a-a0b0-4c9814889b40","impersonationUseSudo":false,"discoveryFailed":false}

In short, I don't see a "children" node or tag in this JSON construct. I wonder if there is an API change between UCD 6.1.x and 6.2.x is the root cause of this issue.

BTW, the "getLowestLevelChildren" method is being called from "getEnvironmentResourceStatusData" method and I see a comment about some "new logic" that was introduced on Dec16/2015. Please check.

If you would like me to step through the code via. Webex let me know, we can schedule some time.

**Comment by shilpsl1** [comment]
Sure we can have a call if @tabladrum is available.

**Comment by spyndi** [comment]
@tabladrum
Were you able to review this thread? I would like to understand the action plan on this UDeploy collector issue. I can assist (if needed) to fix this issue. 

Thank you

**Comment by shilpsl1** [comment]
Hi,
Anyone has any update regarding this.
I even downgraded the UCD to  6.1.3.0 even then I am not getting any dashboard view :( 
Please let me know if anyone has go solution. Thank you!!!

**Comment by satishc1** [comment]
@spyndi @tabladrum 
In DefaultUDeployClient getLowestLevelChildren needs to be changed to handle the 6.2 .  It might be a new api call to Udeploy to get the children based on "hasChildren": true . It needs to be backward compatible with 6.1 also.

The mapping we are looking for is for each childObject:
version
fileName

The images below show the actual uDeploy widget. The rows in the uDeploy widget are links for the various environments. The images for each link when you click on it is also shown.

![image](https://cloud.githubusercontent.com/assets/21199084/23276347/f0bce6d4-f9d7-11e6-87d4-9fa2e2f390be.png)

![image](https://cloud.githubusercontent.com/assets/21199084/23276357/f7f3cf80-f9d7-11e6-884e-01050e51d9a6.png)

![image](https://cloud.githubusercontent.com/assets/21199084/23276372/fe5ff588-f9d7-11e6-95dc-6a619f4689a2.png)

**Comment by shilpsl1** [comment]
@satishc1 
Are the above screen shots for 6.2 or 6.1 udeploy?
what does  It needs to be backward compatible with 6.1 also mean ?

**Comment by tabladrum** [comment]
@shilpsl1: @spyndi and we are collaborating to update the collector and make it compatible to both 6.2 and 6.1. Stay tuned.

**Comment by shilpsl1** [comment]
That's great!!! Thank you :+1:

**Comment by jlk1975** [comment]
Hello all - just curious if anyone knows how soon there might be an update to the collector to make it compatible to both 6.2 and 6.1 uDeploy? Any info on that much appreciated.

**Comment by gokulknewt** [comment]
Hi tabladrum,

We are trying to connect and get the data from IBM UrbanCode Deploy tool(version 6.2.2.0.830952) via Hygieia code. we are able to get the application detail from configuration widget from dropdown. Once hit the save button from widget,In UCD not showing up in Hygieia Dashboard.
Could you please help us out.

FYI,

Please find the image url


![image](https://cloud.githubusercontent.com/assets/23208312/24138629/fa6e17fe-0e3e-11e7-93bf-685862495c6d.png)

**Comment by spyndi** [comment]
@jlk1975 - I have started working on updating Hygieia uDeploy collector to use uDeploy API 6.2.x. Because of other priorities/delays on my side, the work/fix is still in progress. 
I'm targeting the fix to be ready for testing by end of this month. (03/31). Let's touch base next week.

**Comment by spyndi** [comment]
@gokulknewt  - You are running into the same issue that is documented in this issue/thread. The Hygieia uDeploy collector needs to be updated to work with uDeploy version 6.2.x.

**Comment by jlk1975** [comment]
@spyndi - Thank you very much for the response and for working towards a resolution!

**Comment by gokulknewt** [comment]
Thanks lot team for your support.Once done kindly let us know.

**Comment by jlk1975** [comment]
@spyndi - Hi, just checking in with you to see if the updates to the Hygieia uDeploy collector to use uDeploy API 6.2.x might be ready for testing anytime soon? Thanks again for all your work on this!

**Comment by gokulknewt** [comment]
We are ready for testing.Once update its done please let us know.

**Comment by gokulknewt** [comment]
Hi spyndi,
Is there any update on UCD  version(6.2.2) code migration?Could you please let us know once if its done..

Regards,
Gokul k

**Comment by gokulknewt** [comment]
@**spyndi** , For UCD 6.2.2 migration Is there any update?

**Comment by dhivaharNewt** [comment]
@spyndi Any update on this issue regarding UCD version 6.2.2 ?

**Comment by Satheesh-b** [comment]
@spyndi I am too facing the same issue. Is there any update? 

Thanks,
Satheesh. B

**Comment by spyndi** [comment]
@jlk1975 @gokulknewt @dhivaharNewt @Satheesh-b 

An update,
There is quite a lot of changes with 6.2 API, so it's getting a little challenging with incorporating these changes in the existing collector and back porting with 6.1 implementation . I have collected some data points for discussion with @satishc1  and @tabladrum, to make sure it fulfills the existing UI flow and requirement.  Will be reviewing with them shortly.

Has anyone looked into UDeploy 6.2 API? I would like to brainstorm some thoughts/ideas on the implementation.

Thank you.

**Comment by spyndi** [comment]
@satishc1 @tabladrum 

Satish/Topo,

Are you available for a conference call to review the UDeploy 6.2 API findings and also to validate the implementation?

Thank you.

**Comment by Satheesh-b** [comment]
@spyndi 

Any update? Will there be any improvement ? 

Thanks,

Satheesh. B

**Comment by git4giri** [comment]
@tabladrum 
is there any update on this?

**Comment by gokulknewt** [comment]
@Spyndi, Have you done the code migration for UCD 6.2.2? Kindly let me know

**Comment by vvigreen981** [comment]
@spyndi, 
I've been looking at UCD 6.2.1 API but I don't see anything obvious and "elegant" that would return inventory and compliance information. 
 
The only thing I found that is close to what we need is ApplicationProcessRequest resource, it returns the deployment information containing deployment status, component version and environment, but the problem is in getting ApplicationProcessRequest ID in the first place. I couldn't find a way to get it.

Did anyone find better solution? Please share your idea(s).

Thanks,
Victor

**Comment by gokulknewt** [comment]
@tabladrum @spyndi
is there any update on this?

**Comment by spyndi** [comment]
@gokulknewt @git4giri @vvigreen981 

Sorry, just got back from my vacation. I don't have any additional updates on the progress.

I'm finding it bit challenging to retrofit UDeploy 6.2.2 API to accommodate the existing Hygieia Udeploy Collector code/process flow. It appears, we may need to completely re-design/re-engineer the Udeploy collector to support Udeploy API 6.2.2.

I have also reached out to Topo @tabladrum  and Satish @satishc1 for some directions as they are the original authors on this collector.

Victor - It looks like you also reviewed the 6.2.2 API. Are you available to discuss the flow/share your findings and brain storm on an approach?

Thank you.

**Comment by Satheesh-b** [comment]
@spyndi Are you able to fix this issue? Is there any update? Please keep us posted.

**Comment by palayoov** [comment]
Looking forward to this update, we are also running version 6.2.4.2.912589 and having the same issue.  Thank you so much for making this available!

**Comment by vvigreen981** [comment]
Hi folks, 
Finally got some time to push the change for Urban Code collector. It works with version 6.2.*. Haven't tested the old versions but I thing it's not compatible.

**Comment by shouebs** [comment]
@vvigreen981 any idea when the changes would be available in the master ?

**Comment by irvingoe** [comment]
@shouebs the changes were made into his fork due to lack of fallback compatibility, therefore I don't think it'll be available on master any time soon.

**Comment by tabladrum** [comment]
I will take a look and see I can merge (functionally) the two

**Comment by ameetsvshah** [comment]
Hi @tabladrum @spyndi Any update on this issue.
Can you please post the fix here.
We are using 2.0.4 release.
Thanks. Waiting for your response.

**Comment by ameetsvshah** [comment]
@vvigreen981 @tabladrum Please merge your code for this UCD fix. We are stuck on this issue.

**Comment by ameetsvshah** [comment]
@vvigreen981 @tabladrum Any update on release of Udeploy fix for v6.2.x. We are stuck on this and Udeploy is the primary tool in our tool stack. We are unable to showcase pipeline view as well.

**Comment by sai1408** [comment]
@tabladrum Hi Topo, We are also awaiting for this UCD collector fix for v6.2.x . Could you please let us know, when can we expect this?

**Comment by Fairozz** [comment]
I am also facing this issue. Please update once the fix is available.

**Comment by Dinnerspy** [comment]
any news on this issue i am facing it as well

**Comment by Sbrenthughes** [comment]
@mohitsinghal83 as soon as it the branch has been updated
